### PR TITLE
CS-1465-Hint-Refinement

### PR DESCRIPTION
**Pull request by CalebSerafin**
## What type of PR is this.
* [x] Bug
* Change
* [x] Enhancement

### What have you changed and why?
Information:
* Dismiss hint will now also fire on mouse key-binds. (It used to only fire the EH on keyboard.)
* Dismiss key will fire on key down **only once**. However, there is a performance *boost* to the new implementation used.
  * Fast Xor edge detector executes every frame instead of a more lengthy check previously used.
* RenderLoop pauses when there is nothing to draw. performance *boost*.
* Hints will auto dismiss after 15-sec. 
* 10-sec fade animation for the footer to visualise how much time left.
  * Cost: hint rendered every 15 frames (was 10 sec). Not too bad as it builds on previous optimisations.
  * It does make it more responsive in other situations.

TL;DR More features; More Performance.

### Please specify which Issue this PR Resolves.
closes [#1464](https://github.com/official-antistasi-community/A3-Antistasi/issues/1464)
closes [#1465](https://github.com/official-antistasi-community/A3-Antistasi/issues/1465)

### Please verify the following and ensure all checks are completed.

2. [x] Have you loaded the mission in LAN host?
3. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [x] No (Although independent testing is optimal)
* Yes 

# How can the changes be tested?
> Basic

Make sure mission boots.
Boot messages overwrite each other without a dismiss option.
At the end of init, the dismiss option is enabled.
Create some messages and cycle through them on a low-end PC. OR
Wait for them to auto dismiss after 15-sec. There should be a 10-sec footer fade animation.

> Advanced

Key presses should dismiss **only one** hint until your finger is raised off the key and then put back onto the key.
Assign wild key-binds to User Action 12.
Smash your keyboard.
See if keybindings still work as expected.

> Hint Samples

```sqf
["FooBar", "Hello World"] call A3A_fnc_customHint;

//Save Notify
        ["Restart Notification", "Please make your way to HQ as a restart will occure soon™<br/><br/>Please make sure you save your stats at the Map.", true] remoteExec ["A3A_fnc_customHint", 0, false];

["Vaya...", "Parece que sus notificaciones importantes se cifraron.<br/><br/>Nadie espera el cifrado español.", false, ["Pictures\Intel\laptop_error.paa",1]] remoteExec ["A3A_fnc_customHint", 0, false];

// Pre-parse FooBar(Hello World) NoMacro
        private _iconXML = parseText "<img color='#ffffff' image='functions\UI\images\logo.paa' align='center' size='2' shadow='1' shadowColor='#000000' />";
        private _separator  = parseText "<br/><img color='#e6b24a' image='functions\UI\images\img_line_ca.paa' align='center' size='0.60' />";
        private _header = parseText "<br/><br/><t size='1.2' color='#e5b348' shadow='1' shadowColor='#000000'>FooBar</t>";
        private _body = parseText "<br/><br/><t size='1' color='#ffffff' shadow='1' shadowColor='#000000'>Hello World</t><br/>";
        FooBarParse = composeText [_iconXML, _header, _separator, _body, _separator];
        ["FooBar", FooBarParse] call A3A_fnc_customHint;

```
> Push your PC's hardware.
https://github.com/CalebSerafin/FrostsBites-Random-Arma-Scripts/blob/master/A3A_HintRainbowUberFHD.sqf

### Zip of PBO is Provided
_`c2f211f` - 19Sep2020_
[Antistasi-TEST-CS-1465-Hint-Refinement-2-3-1.Altis.pbo.zip](https://github.com/official-antistasi-community/A3-Antistasi/files/5248690/Antistasi-TEST-CS-1465-Hint-Refinement-2-3-1.Altis.pbo.zip)

### CODE REVIEW

**Comment by jaj22** [comment]
Just to check I'm following the logic correctly on the keydown/keyup problem:

1. Arma does not generate events for inputActions, only for keypresses.
2. actionKeys is broken for anything except simple scancodes, so you can't use that to translate.
3. inputAction returns 1 on keyUp for that action, so that doesn't work either.

I don't really understand how the Arma API got through a decade without adding events for inputActions, but that does appear to be the case, so it follows that polling is required.

**Review by jaj22 — APPROVED** [review]
Not too convinced about the footer fade but whatever. Everything seems work reasonably. The code looks solid.

File: A3-Antistasi/functions/UI/fn_customHintInit.sqf [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            A3A_customHint_DismissKeyDown = true;
+            [] call A3A_fnc_customHintDismiss;
+        };
+        false;
+    };
+    waitUntil {!isNull (findDisplay 46)};
+    private _UIDisplay = findDisplay 46;
+    _UIDisplay displayAddEventHandler ["KeyDown", _onKeyDown];
+    _UIDisplay displayAddEventHandler ["MouseButtonDown", _onKeyDown];
+};
```
[the remark this answers is not in the document]
Comment (CalebSerafin) [reply A3-Antistasi/functions/UI/fn_customHintInit.sqf:78]: That's a good point. I'll swap out BIS_fnc_loop for an onEachFrame event handler. Yeah, I can scrap the keyDown event handlers, as it's already going to be running code each frame when someone walks forward ect, so not much performance saving by waiting for key presses.

[the remark this answers is not in the document]
Comment (CalebSerafin) [reply A3-Antistasi/functions/UI/fn_customHintInit.sqf]: I benchmarked the new implementation i chose compared to a more easily readable alternative to decide whether it's worth the trade off.

## Results
A global `inputActionUser12` variable was used instead of `inputAction "User12"` allow testing.
0 or non-zero is value of inputActionUser12.
`[] call A3A_fnc_customHintDismiss;` is replaced by `false;` in each method. This is only called on the rising edge.
The two global variables are reset every loop of the performance test.
Values are in milliseconds.

| Truth Table   | false 0 |  true 0 | false 1 |  true 1 |
|---------------|---------|---------|---------|---------|
| Edge Detector | 0.0030  | 0.0050  | 0.0053  | 0.0030  |
| Conventional  | 0.0050  | 0.0045  | 0.0060  | 0.0050  | 

Settings the global variables by themselves takes 0.0013 ms.
Therefore the adjusted tables compares the methods solely:

| Truth Table   | false 0 |  true 0 | false 1 |  true 1 |
|---------------|---------|---------|---------|---------|
| Edge Detector | 0.0017  | 0.0037  | 0.0040  | 0.0017  |
| Conventional  | 0.0037  | 0.0032  | 0.0047  | 0.0037  | 
|     Ratio     | 2.1765  | 0.8649  | 1.175   | 2.1765  | 

Reacting to edges only happens twice per correct key press.
Therefore the time taken for this operation is negligible.
However, waiting for an edge to appear will be most common: When the user has not pressed the key, when the user's finger is on the key.
Therefore, we see a ~2.2 times performance boost in using the edge detector method. Consuming 0.0017ms of calculations per frame.

Re adding the evaluation time of `inputAction "User12"` (0.0017 ms) we see the total execution time rise to **0.0034ms**. 

### Key down benchmark
```sqf
//  Edge Detector
A3A_customHint_DismissKeyDown = true;
inputActionUser12 = 1;
if ((inputActionUser12 isEqualTo 0) isEqualTo A3A_customHint_DismissKeyDown) then {
    A3A_customHint_DismissKeyDown = !A3A_customHint_DismissKeyDown;
    if (A3A_customHint_DismissKeyDown) then { false; };
};

//  Conventional
A3A_customHint_DismissKeyDown = true;
inputActionUser12 = 1;
if (A3A_customHint_DismissKeyDown && {inputActionUser12 isEqualTo 0}) then {
    A3A_customHint_DismissKeyDown = false;
} else {
    if !(A3A_customHint_DismissKeyDown || {inputActionUser12 isEqualTo 0}) then {
        A3A_customHint_DismissKeyDown = true;
        false;
    };
};
```

**Review by jaj22 — APPROVED** [review]
Minus 20 line commit. That's definitely my favourite type.
